### DISCUSSION TITLE
company-clang doesn't complete on reserved keywords

### DISCUSSION THREAD

**Discussion by wanghaiqiangk** [company-mode/company-mode#1365] [Q&A] [answered]
I know that company-keywords is used for code completion of keywords. However, I just feel strange that company-clang cannot do that. For example, if I type in "ret", or " conti", what I expect is candidate of "return" or "continue" directly appears in the buffer or is added into completion list.

This also results in a nuisance. Company will fail to process the next available backend after company-clang. I have to manually invoke `company-other-backend` or use `company-try-hard`.

Did I miss something? Is this a desirable behaviour?

---

Diagnostic information from `company-diag`.

```plain
Emacs 28.1 (x86_64-pc-linux-gnu) of 2022-04-20 on wang
Company 0.9.13

company-backends: (company-c-headers company-bbdb company-cmake company-capf company-clang company-files
                   (company-dabbrev-code company-gtags company-etags company-keywords)
                   company-oddmuse company-dabbrev)

Used backend: company-clang
Major mode: c++-mode
Prefix: #("ret" 0 1
  (fontified t)
  1 2
  (fontified t)
  2 3
  (fontified t))

Completions: none
```

**Comment by yugaego** [comment, marked as answer]
> Company will fail to process the next available backend after company-clang.

This is a known limitation (a feature).
You could try [grouping](https://company-mode.github.io/manual/Backends.html#Grouped-Backends) the mentioned backends like so:

```elisp
(setq-local company-backends '((company-clang company-keywords)))
```
 
<br>  

> if I type in "ret", or " conti", what I expect is candidate of "return" or "continue" directly appears in the buffer or is added into completion list.

You can also check contents of the`*clang-output*` buffer when testing completions.
For `"re"`, for instance, it may contain
```
...
COMPLETION: restrict
COMPLETION: Pattern : return <#expression#>;
COMPLETION: rewind : [#void#]rewind(<#FILE *#>)
```

Note that `Pattern` support is [marked as a TODO](https://github.com/company-mode/company-mode/blob/93b0284e5a6be8506d194aab00fb613be0f364fd/company-clang.el#L122).

**Reply by dgutov** [reply]
`company-clang` already did show some keywords in completions (such as `const` or `int`), but some - did not.

Thanks for the reminder, I've added support for "patterns" just now. There's still some space for improvement remaining, but the new addition seems useful already.

Not sure if it covers all the useful keywords for a given context, though -- feedback welcome.

**Reply by wanghaiqiangk** [reply]
Thanks for all replies. The support for patterns is just in time. Will try it soon.
